File: src/main/kotlin/com/huawei/cangjie/psi/Call.java
package com.huawei.cangjie.psi;

import com.huawei.cangjie.lexer.CjTokens;

import com.huawei.cangjie.utils.ReadOnly;
import com.intellij.lang.ASTNode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;
public interface Call {

    // SAFE_ACCESS or DOT or so
    @Nullable
    ASTNode getCallOperationNode();

    default boolean isSemanticallyEquivalentToSafeCall() {
        return getCallOperationNode() != null && getCallOperationNode().getElementType() == CjTokens.SAFE_ACCESS;
    }


    @Nullable
    CjExpression getCalleeExpression();

    @Nullable
    CjValueArgumentList getValueArgumentList();

    @ReadOnly
    @NotNull
    List<? extends ValueArgument> getValueArguments();

    @ReadOnly
    @NotNull
    List<? extends LambdaArgument> getFunctionLiteralArguments();

    @ReadOnly
    @NotNull
    List<CjTypeProjection> getTypeArguments();

    @Nullable
    CjTypeArgumentList getTypeArgumentList();

    @NotNull
    CjElement getCallElement();

    enum CallType {
        DEFAULT, ARRAY_GET_METHOD, ARRAY_SET_METHOD, INVOKE, CONTAINS
    }

    @NotNull
    Call.CallType getCallType();
}

File: src/main/kotlin/com/huawei/cangjie/psi/Call.kt
package com.huawei.cangjie.psi

import com.huawei.cangjie.lexer.CjTokens
import com.huawei.cangjie.resolve.scopes.receivers.Receiver
import com.huawei.cangjie.resolve.scopes.receivers.ReceiverValue
import com.huawei.cangjie.utils.ReadOnly
import com.intellij.lang.ASTNode
import kotlin.reflect.KMutableProperty
import kotlin.reflect.KProperty

interface Call {
    // SAFE_ACCESS or DOT or so
    var  noValueArgument: Boolean

    //控制typeArgumentList返回空值
    var noTypeParameter: Boolean
    val callOperationNode: ASTNode?

    val isSemanticallyEquivalentToSafeCall: Boolean
        get() = callOperationNode != null && callOperationNode!!.elementType === CjTokens.SAFE_ACCESS


    val explicitReceiver: Receiver?


    val dispatchReceiver: ReceiverValue?


    val calleeExpression: CjExpression?


    val valueArgumentList: CjValueArgumentList?


    @get:ReadOnly
    val valueArguments: List<ValueArgument>


    @get:ReadOnly
    val functionLiteralArguments: List<LambdaArgument>


    @get:ReadOnly
    val typeArguments: List<CjTypeProjection>


    val typeArgumentList: CjTypeArgumentList?


    val callElement: CjElement

    enum class CallType {
        DEFAULT, ARRAY_GET_METHOD, ARRAY_SET_METHOD, INVOKE, CONTAINS
    }


    val callType: CallType
}
